src/main/java/org/zalando/fahrschein/ManagedCursorManager.java

```diff
@@ -24,49 +24,109 @@
+    }
+
+    @Override
+    public void addStreamId(Subscription subscription, String streamId) {
+        final String eventName = Iterables.getOnlyElement(subscription.getEventTypes());
+
+        LOG.debug("Adding stream id [{}] for subscription [{}] to event [{}]", streamId, subscription.getId(), eventName);
+
+        streams.get(eventName).setStreamId(streamId);
     }
 
     @Override
     public void onSuccess(String eventName, Cursor cursor) throws IOException {
-        final Subscription subscription = subscriptions.get(eventName);
-        final URI subscriptionUrl = baseUri.resolve(String.format("/subscriptions/%s/cursors", subscription.getId()));
 
-        final ClientHttpRequest request = clientHttpRequestFactory.createRequest(subscriptionUrl, HttpMethod.PUT);
+        final SubscriptionStream stream = streams.get(eventName);
+        final String subscriptionId = stream.getSubscriptionId();
+        final URI subscriptionUrl = baseUri.resolve(String.format("/subscriptions/%s/cursors", subscriptionId));
+
+        LOG.debug("Committing cursors for subscription [{}] to event [{}] in partition [{}] with offset [{}]", subscriptionId, stream.getEventName(), cursor.getPartition(), cursor.getOffset());
+
+        final ClientHttpRequest request = clientHttpRequestFactory.createRequest(subscriptionUrl, HttpMethod.POST);
 
         request.getHeaders().setContentType(MediaType.APPLICATION_JSON);
+        request.getHeaders().put("X-Nakadi-StreamId", singletonList(stream.getStreamId()));
 
         try (OutputStream os = request.getBody()) {
             objectMapper.writeValue(os, Collections.singleton(cursor));
         }
 
         try (final ClientHttpResponse response = request.execute()) {
 
-            if (response.getStatusCode().value() == HttpStatus.NO_CONTENT.value()) {
-                LOG.warn("Cursor for event [{}] in partition [{}] with offset [{}] was already committed", eventName, cursor.getPartition(), cursor.getOffset());
-            } else if (response.getStatusCode().is2xxSuccessful()) {
-                LOG.debug("Successfully committed cursor for event [{}] in partition [{}] with offset [{}]", eventName, cursor.getPartition(), cursor.getOffset());
+            final int status = response.getStatusCode().value();
+            if (status == 204) {
+                LOG.debug("Successfully committed cursor for subscription [{}] to event [{}] in partition [{}] with offset [{}]", subscriptionId, eventName, cursor.getPartition(), cursor.getOffset());
+            } else if (status == 200) {
+                LOG.warn("Cursor for subscription [{}] to event [{}] in partition [{}] with offset [{}] was already committed", subscriptionId, eventName, cursor.getPartition(), cursor.getOffset());
+            } else {
+                LOG.warn("Unexpected status code [{}] for subscription [{}] to event [{}] in partition [{}] with offset [{}]", status, subscriptionId, eventName, cursor.getPartition(), cursor.getOffset());
```

Comment:
It should still be a succesful response, codes >= 400 are handled by `ProblemHandlingClientHttpRequest`. Not sure what would happen with the 3xx range though.